solver/llbsolver/file/backend.go

```diff
@@ -205,38 +205,37 @@ func docopy(ctx context.Context, src, dest string, action pb.FileActionCopy, u *
 		copy.WithXAttrErrorHandler(xattrErrorHandler),
 	}
 
+	var m []string
 	if !action.AllowWildcard {
-		if action.AttemptUnpackDockerCompatibility {
-			if ok, err := unpack(ctx, src, srcPath, dest, destPath, ch, timestampToTime(action.Timestamp)); err != nil {
-				return err
-			} else if ok {
-				return nil
-			}
+		m = append(m, srcPath)
+	} else {
+		var err error
+		m, err = copy.ResolveWildcards(src, srcPath, action.FollowSymlink)
+		if err != nil {
+			return err
 		}
-		return copy.Copy(ctx, src, srcPath, dest, destPath, opt...)
-	}
 
-	m, err := copy.ResolveWildcards(src, srcPath, action.FollowSymlink)
-	if err != nil {
-		return err
-	}
-
-	if len(m) == 0 {
-		if action.AllowEmptyWildcard {
-			return nil
+		if len(m) == 0 {
+			if action.AllowEmptyWildcard {
+				return nil
+			}
+			return errors.Errorf("%s not found", srcPath)
 		}
-		return errors.Errorf("%s not found", srcPath)
 	}
 
 	for _, s := range m {
+		d := destPath
 		if action.AttemptUnpackDockerCompatibility {
-			if ok, err := unpack(ctx, src, s, dest, destPath, ch, timestampToTime(action.Timestamp)); err != nil {
+			if ok, err := unpack(ctx, src, s, dest, d, ch, timestampToTime(action.Timestamp)); err != nil {
 				return err
 			} else if ok {
 				continue
 			}
 		}
-		if err := copy.Copy(ctx, src, s, dest, destPath, opt...); err != nil {
+		if action.Parents {
+			d = filepath.Join("/", d, s)
```

Comment:
I thought this would use `IncludePatterns` from https://pkg.go.dev/github.com/tonistiigi/fsutil@v0.0.0-20221114235510-0127568185cf/copy#CopyInfo . Changing the destination directory would mean that the metadata for parent dirs would be the default and not come from the source.

Reply:
Great catch. Working on it.

Reply:
I'm a bit confused. It seems, that `IncludePatterns` is irrelevant here, due to `--chown` feature. Am I'm missing something?

Please take a look at this sequence:

```shell
$ touch $TMP_DIR/src/dir2/subdir2/file2-1.go
$ touch $TMP_DIR/src/dir2/subdir2/file2-2.js
$ touch $TMP_DIR/src/dir2/subdir2/file2-3.go
```

```dockerfile
FROM alpine
RUN adduser -D user
USER user
COPY --chown=user:user --parents ./src/*/*/*.go ./dst2/
```

```shell
$ docker container export parents_test | tar tvf - | grep "dst2"
src/
src/dir2/
src/dir2/subdir2/
src/dir2/subdir2/file2-1.go
src/dir2/subdir2/file2-3.go
drwxr-xr-x 1000/1000         0 2022-12-15 12:33 dst2/
drwxr-xr-x 1000/1000         0 2022-12-15 12:33 dst2/src/
drwxr-xr-x 1000/1000         0 2022-12-15 12:33 dst2/src/dir2/
drwxr-xr-x 1000/1000         0 2022-12-15 12:33 dst2/src/dir2/subdir2/
-rw-r--r-- 1000/1000         0 2022-12-15 12:33 dst2/src/dir2/subdir2/file2-1.go
-rw-r--r-- 1000/1000         0 2022-12-15 12:33 dst2/src/dir2/subdir2/file2-3.go
```

Reply:
Got it - it doesn't respect permissions and such, like this: `chmod g-x $TMP_DIR/src/dir2`.

Reply:
Approach works - very hacky splinter, but works. I'll add some tests to cover it a bit later, then push.